File: pset6/credit/credit.py
from dataclasses import dataclass
import sys
import math
import re


# https://stackoverflow.com/questions/35988/c-like-structures-in-python
@dataclass
class credit_card:
    credit_card_number: str
    numberArray: list
    numberOfDigits: int


def digit_sum(num):
    sum = 0
    while num:
        sum += (num % 10)
        num = num // 10
    return sum


def isAMEX(numberArray: list, numberOfDigits: int):
    return (numberOfDigits == 15) and (numberArray[0]*10 + numberArray[1] in [34, 37])


def isMASTERCARD(numberArray: list, numberOfDigits: int):
    return (numberOfDigits == 16) and (51 <= numberArray[0]*10 + numberArray[1] <= 55)


def isVISA(numberArray: list, numberOfDigits: int):
    return (numberOfDigits in [13, 16]) and (numberArray[0] == 4)


def check_credit_card_type(c: credit_card):
    """ Luhn's Algorithm """

    # Step 1
    sum = 0
    for i in range((c.numberOfDigits - 1) - 1, -1, -2):
        sum += digit_sum(c.numberArray[i] * 2)

    # Step 2
    for i in range(c.numberOfDigits - 1, -1, -2):
        sum += c.numberArray[i]

    # Step 3
    lastDigit = sum % 10
    if (lastDigit != 0):
        print("INVALID")
    else:
        startNumbers = c.numberArray[0] * 10 + c.numberArray[1]
        if isAMEX(c.numberArray, c.numberOfDigits):
            print("AMEX")
        elif isMASTERCARD(c.numberArray, c.numberOfDigits):
            print("MASTERCARD")
        elif isVISA(c.numberArray, c.numberOfDigits):
            print("VISA")
        else:
            print("INVALID")


def main():
    while (True):
        credit_card_number = input("Number: ")
        if (not re.match("^(\d{13}|\d{15}|\d{16})$", str(credit_card_number))):   # regular expression (\d = [0-9])
            print("INVALID")
            sys.exit(0)   # Or: return
        else:
            numberArray = [int(digit) for digit in credit_card_number]   # list comprehension
            numberOfDigits = math.floor(math.log10(int(credit_card_number))) + 1
            c = credit_card(credit_card_number, numberArray, numberOfDigits)
            break

    check_credit_card_type(c)


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <string.h>


// Prototypes
void toArray(int arr[], long num);
int digit_sum(long num);
bool isAMEX(const int *numberArray, const int numberOfDigits);
bool isMASTERCARD(const int *numberArray, const int numberOfDigits);
bool isVISA(const int *numberArray, const int numberOfDigits);

int main(void)
{
    long credit_card_number;
    int numberOfDigits;
    bool isValid = false;
    while (!isValid)
    {
        credit_card_number = get_long("Number: ");
        numberOfDigits = floor(log10(credit_card_number)) + 1;
        if (numberOfDigits != 13 && numberOfDigits != 15 && numberOfDigits != 16)
        {
            printf("INVALID\n");
            return 0;
        }
        else
        {
            isValid = true;
        }
    }

    int numberArray[100];
    toArray(numberArray, credit_card_number);

    /* Luhn's Algorithm */

    // Step 1
    int sum = 0;
    for (int i = (numberOfDigits - 1) - 1; i >= 0; i -= 2)
    {
        sum += digit_sum(numberArray[i] * 2);
    }

    // Step 2
    for (int i = (numberOfDigits - 1); i >= 0; i -= 2)
    {
        sum += numberArray[i];
    }

    // Step 3
    int lastDigit = sum % 10;
    if (lastDigit != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        //int startNumbers = numberArray[0] * 10 + numberArray[1];
        if (isAMEX(numberArray, numberOfDigits))
            //if ((numberOfDigits == 15) && (startNumbers == 34 || startNumbers == 37))
        {
            printf("AMEX\n");
        }
        else if (isMASTERCARD(numberArray, numberOfDigits))
            //else if ((numberOfDigits == 16) && (startNumbers >= 51 && startNumbers <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if (isVISA(numberArray, numberOfDigits))
            //else if ((numberOfDigits == 13 || numberOfDigits == 16) && (numberArray[0] == 4))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    return 0;
}


void toArray(int arr[], long num)
{
    int numberOfDigits = floor(log10(num)) + 1;
    for (int i = 0; i < numberOfDigits; ++i)
    {
        arr[(numberOfDigits - 1) - i] = num % 10;
        num /= 10;
    }
}


int digit_sum(long num)
{
    int sum = 0;
    while (num)
    {
        sum += (num % 10);
        num = num / 10;
    }
    return sum;
}

bool isAMEX(const int *numberArray, const int numberOfDigits)
{
    return (numberOfDigits == 15) && strchr("3", numberArray[0] + '0') && strchr("47", numberArray[1] + '0');
}

bool isMASTERCARD(const int *numberArray, const int numberOfDigits)
{
    return (numberOfDigits == 16) && strchr("5", numberArray[0] + '0') && strchr("12345", numberArray[1] + '0');
}

bool isVISA(const int *numberArray, const int numberOfDigits)
{
    return (numberOfDigits == 13 || numberOfDigits == 16) && strchr("4", numberArray[0] + '0');
}